use initial value of inputs without discarding them

seanhess opened the pull request:
This changes start-app so that it doesn't discard the first value of input signals when it initializes. Personally, I needed this for routing with [`elm-history`](https://github.com/TheSeamau5/elm-history/), but it has been requested multiple times. 

See #35, #19, and #21. 

I'm more than happy to clean this up, or let me know how the approach should change. I couldn't tell whether this wasn't being worked on because you guys are busy, or because it was a bad idea for some reason.

Either way, this is immediately useful to me, and I think others, so I'm hoping we can get it in. Let me know what I should change first!

seanhess committed:
use initial value of inputs without discarding them

jvoigtlaender commented:
I think you changed the semantics in a harmful way. Do we agree that for people that don't want/need to use the initial values, `start-app` after the change should behave like `start-app` before the change? If so, your pull request violates this. The reason is that `Signal.mergeMany` and `Signal.Extra.mapMany List.concat` have _very_ different behavior.

jvoigtlaender commented:
Let me expand just a bit. Basically you should be able to see misbehavior with your "fixed" version of `start-app` for any app that has at least two signal sources in `config.inputs`.

Say you have something based on `Mouse.position` and something based on `Keyboard.whatever` in there. With the original version of the code, when `Mouse.position` has an event, the corresponding `Action` will be processed, and when `Keyboard.whatever` has an event, the `Action` corresponding to that is processed. With your version of the code, when `Mouse.position` has an event, the corresponding `Action` will be processed, but _in addition_ the last `Action` that came from `Keyboard.whatever` will also be processed _again_. Try it out.

seanhess commented:
Ok that makes sense. I'll have to think about it more. Do you have an idea
of the direction to head from here?

On Tuesday, October 20, 2015, Janis Voigtländer notifications@github.com
wrote:

> Let me expand just a bit. Basically you should be able to see misbehavior
> with your "fixed" version of start-app for any app that has at least two
> signal sources in config.inputs.
> 
> Say you have something based on Mouse.position and something based on
> Keyboard.whatever in there. With the original version of the code, when
> Mouse.position has an event, the corresponding Action will be processed,
> and when Keyboard.whatever has an event, the Action corresponding to that
> is processed. With your version of the code, when Mouse.position has an
> event, the corresponding Action will be processed, but _in addition_ the
> last Action that came from Keyboard.whatever will also be processed
> _again_. Try it out.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/evancz/start-app/pull/37#issuecomment-149714357.

jvoigtlaender commented:
To fix the immediate problem I pointed out, replace `Signal.mergeMany` (or your `Signal.Extra.mapMany List.concat`) by the following function:

``` elm
fun : List (Signal (List Action)) -> Signal (List Action)
fun signalList =
  case List.reverse signalList of
    [] ->
        Debug.crash "We know this cannot happen!"

    signal :: signals ->
        List.foldl (Signal.Extra.fairMerge List.append) signal signals
```

But even after such a fix, I would be opposed to your change. The API is not expressive enough. Assume my app should be able to react to mouse clicks, so I put some action signal derived from `Mouse.clicks` into my `config.inputs`. With your changes, the app will behave as if the mouse is always clicked once at the beginning of the program (because you interpret the initial value of `Mouse.clicks : Signal ()` as being an actual mouse click that needs to be handled by the `update`-function).

seanhess committed:
added inits: signals that are processed when the app loads

seanhess commented:
Ok, I've integrated your fix, but more importantly, I changed the API to be expressive enough to handle both cases. @jvoigtlaender Could you review the diff now? 

We may want to change the names of things, but it should have the power to express everything. I added a new field to config

```
type alias Config model action =
    { ...
    , inits : List (Signal.Signal action)
    }
```

It works just like inputs, but any signals you put in there will be applied to update when the application loads, while signals you put in inputs work exactly the same as before. 

For example:

```
routes : Signal Action
routes = Signal.map Route History.paths

clicks : Signal Action
clicks = Signal.map (always ClickTest) Mouse.clicks

// this will run a Route action when it loads, but not ClickTest until you click once
config : Config Model Action
config =
  { init = init
  , update = update
  , view = view
  , inits = [routes]
  , inputs = [clicks]
  }

// This will run ClickTest and Route on load
config : Config Model Action
config =
  { init = init
  , update = update
  , view = view
  , inits = [routes, clicks]
  , inputs = []
  }

// This will run nothing on load... only once you click or change the url path
config : Config Model Action
config =
  { init = init
  , update = update
  , view = view
  , inits = []
  , inputs = [routes, clicks]
  }
```

Does this resolve your concerns? How should we rename things to make it more clear?

jvoigtlaender commented:
I think that specific `Signal.mergeMany` which is now back in your code should also be rephrased using `Signal.Extra.fairMerge`.

That's a concern that's actually independent of the "intial values handling" thing (see https://github.com/evancz/start-app/issues/38), but fixing it here as well would make it easier to review the net result of having initial value handling as well, as you request here.

Apart from that concern, two brief points:
- I have no strong opinion about names (usually, and here as well).
- I have my doubts that Evan will accept this pull request as long as it depends on `elm-signal-extra`. (Which, on the other hand, it has to, because relevant functionality is missing from `core`.)

seanhess commented:
I would be happy to continue improving this, but not if it's never going to see the light of day. Can we figure this out?

Maybe I just don't understand elm's culture, but why not let start-app depend on elm-signal-extra? I mean, I understand the sentiment that core stuff should have few dependencies, but that's clearly impractical in this case. Or the other solution is to merge some of elm-signal-extra into core. Or I could see delaying temporarily if 0.16 is going to change everything or something. 

For the 3rd party module ecosystem to be healthy, we need to be able to build off others work, right?

I guess I'm asking how the community handles this sort of thing. I don't have enough visibility to know what to do, or how this could be resolved.

jvoigtlaender commented:
I'm afraid I don't have the answers. Also my last comment above wasn't from an authorative perspective. So I may be wrong, Evan may actually be willing to let `start-app` depend on `elm-signal-extra`. So in the end, you probably have to hope for @evancz to weigh in. But you shouldn't be too surprised if his weighing in will be a reprise of [his comment](https://github.com/evancz/start-app/issues/19#issuecomment-130388039) I pointed you to before.

seanhess committed:
use fairMerge instead of merge. Should resolve #38

seanhess commented:
@evancz Can you give me some guidance here? 

@jvoigtlaender If Evan isn't willing to merge, I could just publish my fork to the package list right? I suppose that's the advantage to fully qualified package names.

I've implemented the recommended change in #38, but I would appreciate some review.

jvoigtlaender commented:
Okay, I'll take a look. Might have to wait till next week, though.

(And yes, you can publish a fork as `seanhess/start-app`.)

simonh1000 commented:
For what it is worth, I have just hit the same issue and for the same reason - I want to be able to process the URL pathname as a page loads - important for deep linking etc...

seanhess commented:
@simonh1000 Would it be useful to you if I published my fork to the package registry?

simonh1000 commented:
Let's see whether the extra interest can trigger a decision by @evancz.

jvoigtlaender commented:
@seanhess, I've looked at your current code version. I don't think you have addressed this concern of mine:

> Assume my app should be able to react to mouse clicks, so I put some action signal derived from `Mouse.clicks` into my `config.inputs`. With your changes, the app will behave as if the mouse is always clicked once at the beginning of the program (because you interpret the initial value of `Mouse.clicks : Signal ()` as being an actual mouse click that needs to be handled by the `update`-function).

No matter whether I put my `Mouse.clicks` into your `config.inputs` or your `config.inits`: Since both of these will flow into the `inputs = combineInputs (mergedInputs :: mergedInits)`, the initial value of which is then processed by `updateStart`, it is still the case that the app will behave as if the mouse is always clicked once at the beginning of the program.

How about you actually do some testing of this case I brought up before?

I also have some comments on possible simplifications of the code, though it doesn't really matter as long as the semantics is not correct. Anyway, here goes: The `List.reverse` in `combineInputs` should not be needed. The API anyway makes no promises about the order of processing actions that arrive at "the same time". Moreover, your code contains only exactly one call of `combineInputs`. That call is with a guaranteedly non-empty list. So it should not be necessary to pattern-match/check for the empty list in `combineInputs`. Instead inlining `combineInputs` would lead to simpler code.

Finally, having thought more about your suggested API now (independently of naming), I think it's not the API I would like to see. It's too unclear and error prone to have these two lists `config.inputs` and `config.inits` that are both of the same type. Programmers would probably often not know what to put where, and will get subtle deviations from what they expect the behavior of their app to be.

The API I would like to see here is one where the types play a role in preventing such issues. Unfortunately, that relies on what is sometimes called "the Signal/Stream idea" on the mailing list, and of which it is uncertain whether it will be realized (it was very close to being realized in one Elm release, then retracted).

jvoigtlaender committed:
refactoring

jvoigtlaender committed:
fix semantics

jvoigtlaender commented:
@seanhess, I've fixed the semantic issue I mention above, with a pull request against your version: https://github.com/seanhess/start-app/pull/1 (but without testing on my side).

My concerns about the API are not addressed by this, but at least it should be a version that works appropriately.

seanhess committed:
Merge pull request #1 from jvoigtlaender/inits

Simplify; Fix semantics

seanhess commented:
I _did_ test your case before I added it here. I must have undone it somehow on a subsequent commit or something. What we really need is a test suite for start-app. 

I'll be the first to admit that I'm still new enough to Elm and its culture to not be sure exactly how to accomplish this, which is why I'm asking for help. I want to see _something_ done though. Thank you for all you've done to improve this. I appreciate you taking action and your insight.

bburdette commented:
Just gave this a try today.  Mine was the 'unable to get initial window size' problem.  Works!  The separate 'inits' and 'inputs' lists make sense to me in this instance at least, that would allow one to wrap the same signal in a different Action if you wanted to distinguish the initial value from a subsequent update.  Hope to see this or another solution mainstreamed soon.

BrianHicks commented:
This is working just fine for me, too. (context: elm-history). +1

statusfailed commented:
+1, would very much like to have this. Currently using an approach by @freakingawesome to work around the problem with `elm-history`:

```
17:10 < chadg> my Main.elm: https://gist.github.com/freakingawesome/f8b9e14e0e058c9bedb2
17:10 < chadg> my javascript initializer: https://gist.github.com/freakingawesome/2fc1b687037442d89e6b
```

kurtharriger commented:
I also ran into this issue and posted about it in the elm-discuss here: 
https://groups.google.com/forum/#!topic/elm-discuss/rsCgT_eR9UU

I think this is perhaps the reason why [baconjs](https://baconjs.github.io/) models `Properties` and `EventStream`s differently.

`Properties` represent an initial value that changes over time. `EventStream`s are just a stream of events without a current value. 

http://stackoverflow.com/questions/29297328/bacon-js-practical-difference-between-a-stream-and-a-property

It makes sense that foldp would need an initial value if the input is an `EventStream` such as `Mouse.clicks`. But if the input is a `Property` the current values could be used to construct the initial state as is done by `foldp'`

`Properties` have a `changes()` method which returns an `EventStream`, effectively ignoring current value.  An `EventStream` can also be converted to a `Property` by providing a default value. I think this would semantically resolve the issue of "I want to only update the model when there is a change to Mouse.position"

In elm `Signals` appear to be `Properties` and `EventStreams` are forced to have an initial state when one does not semantically exist in practice. In my experience so far this forces creation of NoOp actions for initial mailbox states and such.

rgrempel commented:
In terms of Elm code that might be a little like `EventStream`, there is:

http://package.elm-lang.org/packages/Apanatshka/elm-signal-extra/5.7.0/Signal-Stream

So, in principle, an alternate way of dealing with the overall question is to use `Signal` in some circumstances and `Stream` in others.

bburdette commented:
It seems like what would be better than having the values from the list of 'inits' signals go to the update function would be to somehow pass their values in to the init function as arguments.  That way you wouldn't have to rely on default values for things like window dimension, which are immediately superseded in the first call of update().  

How to actually do this is a bit of a trick though.  Perhaps if the init is required to take a tuple of values corresponding to the signals specified in the 'inits' (which could also be a tuple?).  Not sure if that's possible in Elm.

seanhess commented:
I thought about doing that but decided update was better because you could
trigger effects too. If you just passed the values in to init you couldn't
do that.

On Monday, November 30, 2015, bburdette notifications@github.com wrote:

> It seems like what would be better than having the values from the list of
> 'inits' signals go to the update function would be to somehow pass their
> values in to the init function as arguments. That way you wouldn't have to
> rely on default values for things like window dimension, which are
> immediately superseded in the first call of update().
> 
> How to actually do this is a bit of a trick though. Perhaps if the init is
> required to take a tuple of values corresponding to the signals specified
> in the 'inits' (which could also be a tuple?). Not sure if that's possible
> in Elm.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/evancz/start-app/pull/37#issuecomment-160658978.

bburdette commented:
But init() can spawn effects because it returns (model, Effects action)... or maybe you mean something different from that?

seanhess commented:
Yeah you're right of course. I guess I only remember the feeling that doing
multiple updates was better but I don't remember why :)

On Monday, November 30, 2015, bburdette notifications@github.com wrote:

> But init() can spawn effects because it returns (model, Effects action)...
> or maybe you mean something different from that?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/evancz/start-app/pull/37#issuecomment-160799738.

danyx23 commented:
I just ran into this problem as well and it took me a while to find this issue. It seems that most people in here ran into this problem with either Window.dimensions or the history package. I think it would be really great if those two modules mentioned this problem and linked to a way to solve it, and at the moment this issue here seems to be the best resource. It certainly would have saved me some time and frustration. Thoughts?

glenjamin commented:
Did this fork ever get published? I've read the thread but I'm not clear what the suggested workaround is at the moment?

seanhess commented:
No, the fork isn't published to the package manager, but it's only one
file. As a workaround, copy the file into your project, and make sure to
add elm-signal-extra to your dependencies.

On Sat, May 7, 2016 at 3:57 PM Glen Mailer notifications@github.com wrote:

> Did this fork ever get published? I've read the thread but I'm not clear
> what the suggested workaround is at the moment?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/evancz/start-app/pull/37#issuecomment-217671076

evancz commented:
0.17

seanhess commented:
@evancz could you link us to how this is solved in 0.17? Thanks!

evancz commented:
Can you ask on the Elm slack channel? Will be more efficient. There exist examples of doing it with window. It'll be the same as that.
